src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java

```diff
@@ -96,4 +100,40 @@ public ProgrammingSubmission notifyPush(Long participationId, Object requestBody
         participationService.save(participation);
         return programmingSubmission;
     }
+
+    /**
+     * A pending submission is one that does not have a result yet and is not older than RESULT_WAIT_LIMIT_SECONDS.
+     *
+     * @param participationId the id of the participation get the latest submission for
+     * @return the latest pending submission if exists or null.
+     * @throws EntityNotFoundException if the participation for the given id can't be found.
+     * @throws IllegalArgumentException if the participation for the given id is not a programming exercise participation.
+     * @throws IllegalAccessException if the user does not have access to the given participation.
+     */
+    public ProgrammingSubmission getLatestPendingSubmission(Long participationId) throws EntityNotFoundException, IllegalArgumentException, IllegalAccessException {
+        Participation participation = participationService.findOne(participationId);
+        if (participation == null) {
+            throw new EntityNotFoundException("Participation with id " + participationId + " could not be retrieved!");
+        }
+        if (!(participation instanceof ProgrammingExerciseParticipation)) {
+            throw new IllegalArgumentException("Participation with id " + participationId + " is not a programming exercise participation!");
+        }
+        if (!programmingExerciseParticipationService.canAccessParticipation((ProgrammingExerciseParticipation) participation)) {
+            throw new IllegalAccessException("Participation with id " + participationId + " can't be accessed by user " + SecurityUtils.getCurrentUserLogin());
+        }
+
+        Optional<ProgrammingSubmission> submissionOpt = programmingSubmissionRepository.findFirstByParticipationIdOrderBySubmissionDateDesc(participationId);
+        if (!submissionOpt.isPresent() || submissionOpt.get().getResult() != null) {
+            // This is not an error case, it is very likely that there is no pending submission for a participation.
+            return null;
+        }
+        ProgrammingSubmission submission = submissionOpt.get();
+        boolean submissionDateIsWithinWaitLimit = ChronoUnit.SECONDS.between(submission.getSubmissionDate(), ZonedDateTime.now()) <= RESULT_WAIT_LIMIT_SECONDS;
+        if (submissionDateIsWithinWaitLimit) {
+            return submission;
+        }
+        else {
+            return null;
```

Comment:
I am not sure about this case. Even if the submission is older than 60s and there is no result, we should probably show this in the client. So we might need to move the time logic into the client, so that we can still show an error message after 60s.

Reply:
This would mean that if the users last submission does not have a result, every time he loads the page the building indicator is shown and every time will disappear after 1 minute.
I think this would be confusing to the user.

Reply:
But what do we display in the client if null is returned? I don’t think we can distinguish between the other case above where null is returned, I.e no submission exists.
I think the client can handle the timeout.
The last pending submission should include the submission date so we can still display the animation of current date - submission date < timeout, or an error message in the other case.
And showing the error message is important so that the user knows that something went wrong

Reply:
You are right, this is an important error case for the client as the student needs to know that he needs to submit again. I changed it accordingly.